### PR TITLE
feat(wallet/ui): interactive, background signing

### LINKED ISSUE #4406: delegate from interactive signer to background signer for low-privilege actions
## What is the Problem Being Solved?

refs: #4398 

The on-chain wallet backend needs explicit approvals for operations that withdraw from purses, but we don't want to require explicit approvals for most user gestures.

## Description of the Design

Our CosmJS integration is intended to be used as follows:

- start with a "warm" or "cold" keypair in the Signer (Keplr/MetaMask)
- UI guides user through an optional preapproval process.  If not preapproved, then the wallet uses the "warm" keypair for every chain-modifying operation tx (with the full Signer approval process)
  - generate a "hot" Agoric CosmJS keypair, store encrypted with user password in browser
  - submit a batch of authorisation messages as one Signer transaction:
    - `feegrant` from the cold account to the hot account
    - `authz` to allow `MsgWalletAction` from hot account on behalf of cold account
    - optional - `authz` to allow `MsgWalletSpendAction` from hot account on behalf of cold account
- if user has a "hot" keypair, they must enter a password once on load to use it instead of the Signer
- in order to withdraw anything from a purse, the eventual send needs to be done via "MsgWalletSpendAction", where the backend basically remaps the marshalled input slot references as:
  - `refs.map(ref => (spendActionAmplifier.get(ref) || ref))`
  - and then handles the eventual send in the same way as `MsgWalletAction`
- categories of restricted "spend" behaviour:
  - sending from a purse to Zoe escrow as part of offer approval
  - sending from a bank purse (Cosmos-level account, `urun`, `ubld`) to anywhere else
  - NOT sending between purses otherwile

## Security Considerations

Needs review by security folks.

### PR DESCRIPTION
**Pull request by dckc**
closes: #4406
refs: #3628; obsoletes #5761

## Description

`makeInteractiveSigner` supports signing offers as in the initial smart wallet swap demo. (#5354)

It also supports authorizing a background signer which we plan to use in other user stories.
In particular, it uses `authz`. (It should use `feegrant` too, but see #5825 as well as https://github.com/cosmos/cosmjs/issues/1155 ) 

### Security Considerations

The secret key to the background signer is stored in `localStorage` in the clear, protected only by the same-origin policy; this risk is moderate, given:
  - the `authz` grant is limited to low-privilege `MsgWalletAction` messages
  - the grant is limited to 4hrs
  - the grant is limited to spending 0.25IST

### Documentation Considerations

End-user documentation for this feature is in order, not to mention completed UI support.

### Testing Considerations

Testing so far is manual. (_integration test built using a temporary button removed on request_)

### CODE REVIEW

**Comment by dckc** [comment]
naming idea: `makeNonspendingKey`

brainstorming with @turadg

File: packages/wallet/ui/src/components/Requests.jsx
Comment (samsiegart) [inline packages/wallet/ui/src/components/Requests.jsx]: I think it's okay to not include this button in the merge, we have the code in other branches to reference

**Review by samsiegart — COMMENTED** [review]
I don't have a grasp of all the stuff in keyManagement.js but the rest looks good to me assuming we're getting keplr to actually pop up with an approval window. Will defer approval to @michaelfig for the rest.

File: packages/wallet/ui/src/components/Requests.jsx
Comment (dckc) [reply packages/wallet/ui/src/components/Requests.jsx]: ok; it's gone

**Comment by dckc** [comment]
chatting with @michaelfig 
> the grant is limited to 4hrs

he suggests aligning with a browser localStorage norm.

Also, maybe @rowgraus  should weigh in? should that be a week?

File: packages/wallet/ui/package.json
Comment (michaelfig) [inline packages/wallet/ui/package.json:11]: Yaay!  Thanks @kriskowal!

File: packages/wallet/ui/src/contexts/Provider.jsx
Comment (michaelfig) [inline packages/wallet/ui/src/contexts/Provider.jsx:5]: Let's (later) introduce a powerful module that closes over these dependencies and is imported by `Provider.jsx` to keep `Provider.jsx` simpler.

It's fine for now.

File: packages/wallet/ui/src/util/chainInfo.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/chainInfo.js:3]: Noice toipes.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
  return harden({
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
  return harden({
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js:224]: I think CosmJS exports some message-specific facilities to make this kind of process simpler.  This is a fine approach until we sync better with the CosmJS team on our use cases.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js:98]: Would be nice to make this portion of the map part of `@agoric/cosmic-proto`.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
export const BROWSER_LOCAL_STORAGE_KEY = 'agoric.eis0Aigi';
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js:180]: Use the same `?.` expression here as elsewhere and throw a nicer error?

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
          spendLimit: [{ denom: 'uist', amount: allowance }],
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
  console.log('InteractiveSigner', { offlineSigner });
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (michaelfig) [inline packages/wallet/ui/src/util/keyManagement.js]:
```suggestion
  console.debug('InteractiveSigner', { signingClient });
```

**Review by michaelfig — APPROVED** [review]
Given fixes, this looks good to merge!

File: packages/wallet/ui/src/util/keyManagement.js
Comment (turadg) [inline packages/wallet/ui/src/util/keyManagement.js]: I don't get this random string. Wouldn't it be better to be human readable as to what's kept inside? e.g. `agoric.wallet.backgroundSignerKey`. 

It could be versioned to signal breaking changes, but I think if backwards compatibility breaks you'd just want to empty it out and start over.

File: packages/wallet/ui/src/contexts/Provider.jsx
Comment (turadg) [inline packages/wallet/ui/src/contexts/Provider.jsx:158]: better to document this with code, which will provide feedback when the assumptions are violated. e.g. down below,
```
assert(fetch, 'Missing window.fetch');
assert(keplr, 'Missing window.keplr');
assert(getBytes, 'Missing Random.getBytes');
```

File: packages/wallet/ui/src/contexts/Provider.jsx
Comment (turadg) [inline packages/wallet/ui/src/contexts/Provider.jsx:240]: we should have an error toast for failures like this. if we don't have a toast component yet, adding one is out of scope but should be scheduled work before launch. cc @samsiegart

File: packages/wallet/ui/src/util/SuggestChain.js
Comment (turadg) [inline packages/wallet/ui/src/util/SuggestChain.js:14]: please type

File: packages/wallet/ui/src/util/SuggestChain.js
Comment (turadg) [inline packages/wallet/ui/src/util/SuggestChain.js]: if `caption` can be undefined then it's not `string`. let's type the function signature instead of inline.

```suggestion
/** @type {{
  * networkConfig: NetworkConfig
  * caption?: string
  * io?: Partial<IoPowers>
  * }}
  */
export async function suggestChain(networkConfig,, caption, io = {}) {
```

Comment (turadg) [reply packages/wallet/ui/src/util/SuggestChain.js]: Though second thought, why make `io` optional? Please make them required so this module does not depend anything ambient. Then you can remove `/* global globalThis */`.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (turadg) [inline packages/wallet/ui/src/util/keyManagement.js]: `getBytes` without the `Random` namespace reads overly generic. please rename to something that conveys that it's generating random bytes. consider `makeRandomBytes`.

that could mean having to map the name at the callsite but I'd also suggest changing this from an option map to positional parameters. then the caller doesn't have to be concerned with the name in this context.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (turadg) [inline packages/wallet/ui/src/util/keyManagement.js]: consider lowering the log level or removing.
```suggestion
    console.debug(
```

File: packages/wallet/ui/src/util/keyManagement.js
Comment (turadg) [inline packages/wallet/ui/src/util/keyManagement.js:188]: yay types!

File: packages/wallet/ui/src/util/keyManagement.js
Comment (turadg) [inline packages/wallet/ui/src/util/keyManagement.js]: should `decoded` contain falsy? if not, filter before returning.

also if it wasn't supposed to, consider defining the return type of `queryGrants` as that would have highlighted the problem.

**Review by turadg — CHANGES_REQUESTED** [review]
Some small changes requested.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (dckc) [reply packages/wallet/ui/src/util/keyManagement.js]: it's not terribly well motivated... sort of a vague fear of colliding with... I don't know what.

Giving it some thought, I suppose it's scoped to our origin. So I just need to be sure I don't collide with any other code we write. But again... `genpass -1` is kind of a quick-and-dirty way to be 99.99999% sure of that :)

The alternative seems to be some sort of central registry agreed by all (potential) contributors.

Or maybe `agoric.wallet.backgroundSignerKey` has as effectively as much entropy as `genpass -1` and I shouldn't worry about collisions?

Comment (turadg) [reply packages/wallet/ui/src/util/keyManagement.js]: > Or maybe agoric.wallet.backgroundSignerKey has as effectively as much entropy as genpass -1 and I shouldn't worry about collisions?

Yes, I believe that to be the case.

File: packages/wallet/ui/src/contexts/Provider.jsx
Comment (samsiegart) [reply packages/wallet/ui/src/contexts/Provider.jsx:240]: Yea we have toasts but I think I need to do some work to make the toast service more robust and reusable across the app

File: packages/wallet/ui/src/util/keyManagement.js
Comment (dckc) [reply packages/wallet/ui/src/util/keyManagement.js]: I'm pretty fond of named parameters for IO powers and have spread the convention pretty far by now :) _IOU docs to institutionalize it, perhaps._

It's also my habit to inherit the names from whatever API they come from, but I'll grant that `getBytes` is pretty worthless in this context. Changing to `io.csprng`.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (dckc) [reply packages/wallet/ui/src/util/keyManagement.js:180]: On further consideration, elsewhere we were dealing with the keplr API, which is doing IO to come up with an address and hence may fail. In this case, we're getting the address from `DirectSecp256k1Wallet`, which is doing only integer calculations that don't have that sort of failure mode. Any failures would show up at the `await wallet.getAccounts()` above.

File: packages/wallet/ui/src/contexts/Provider.jsx
Comment (dckc) [reply packages/wallet/ui/src/contexts/Provider.jsx:158]: `assert`s for fetch and keplr make sense, but `Missing Random.getBytes` would be a static error.
`Random.getBytes` is still noteworthy because it is ambient authority, contrary to [OCap Discipline](https://github.com/Agoric/agoric-sdk/wiki/OCap-Discipline).

File: packages/wallet/ui/src/contexts/Provider.jsx
Comment (dckc) [reply packages/wallet/ui/src/contexts/Provider.jsx:240]: I'm adding a TODO.

File: packages/wallet/ui/src/util/SuggestChain.js
Comment (dckc) [reply packages/wallet/ui/src/util/SuggestChain.js]: Why indeed. It looks like we addressed the only call site in this regard.

File: packages/wallet/ui/src/util/keyManagement.js
Comment (turadg) [inline packages/wallet/ui/src/util/keyManagement.js]: a little ugly but I see it's necessary. might want to reference `UNTIL https://github.com/microsoft/TypeScript/issues/16655` so we know when we can remove this

**Review by turadg — APPROVED** [review]

File: packages/wallet/ui/src/util/keyManagement.js
Comment (dckc) [reply packages/wallet/ui/src/util/keyManagement.js]: oh. JS Arrays seem to have `flatMap` now... i.e. map and filter in one...

**Pull request merged**
